modules/ROOT/pages/dataweave-type-system.adoc

```diff
@@ -0,0 +1,289 @@
+
+A type system defines a set of constraints to a set of constructs such as:
+
+* Variables
+* Function parameters
+
+These constraints are used in the type checking phase where DataWeave ensures that the values assigned
+to variables or the arguments for a function call respects its constraints.
+
+For example:
+
+----
+var userName: String = "John"
+----
+
+The variable definition for `userName` has the constraint of assigning a value of type String.
+If some other type value is assigned, a type checking error is raised.
+
+----
+fun toUser(id: Number, userName: String): String = "you called the function toUser!"
+----
+
+The constraints to call the function `toUser` are that the first argument has to be of type Number, the second one of type String (or to be coercible to those types). Another constraint here is that the result of the function has to be of type String.(this constraint isn't used in the call for `toUser`).
+
+----
+toUser(123, "John")
+toUser("123", true)
+----
+
+Though you would expect the second call to fail, autocoercions allow it to be successful and both of them work. In the case of the second function call, there is an autocoercion of the String `"123"` to the Number value `123`, and the Boolean `true` to the String value `"true"` to being able to call the function. +
+An example of a call that throws a type checking error is `toUser("a 12", "John")` because the String `"a 12"` isn't coercible to a Number. +
+
+DataWeave also uses a global type inference algorithm to validate your code even if no types are specified as
+constraints.
+
+Note that the use of constraints is optional but this can be useful in big scripts or multiple scripts where the type system could help you avoid bugs in your DataWeave logic by running the type checking algorithm.
+
+The types supported by the DataWeave type system are divided into three categories:
+
+* Simple Types
+* Composite Types
+* Complex Types
+
+== Simple Types
+
+Simple types represent values such as string, Booleans, etc. These values are atomic, they are not composed by other values. These are the simple types:
+
+* String
+* Boolean
+* Number
+* Regex
+* Null
+* Temporal: Date, DateTime, LocalDateTime, LocalTime, Time, Period
+//Key, KeyValuePairType, NameType, NameValuePairType, NamespaceType, RangeType , ReferenceType, TypeParameter, TypeType, UriType
+
+=== Null
+
+Null is a type of only one value, the value `null`, which means that this value `null` cannot be assigned to the type String or any other type except for the type Null.
+
+For example, the repeat function in the String Module has the signature `repeat(String, Number): String`, which means that it only accepts as a first parameter a value of type String, and as a second parameter a value of type Number. The function returns a value of type String.
+
+----
```

Comment:
```suggestion
.DataWeave Script:
----
%dw 2.0
import * from dw::core::Strings
output application/json
```
Not sure what output format or MIME type we want to use since we're not showing the actual output data, but we need to provide output and import directives to avoid throwing an error in Studio.

Also, I'm actually seeing `.Transform` rather than `.DataWeave Script:` elsewhere in this file, but I think the latter might be even more clear than the former.

Reply:
I'd do something like the above on all the examples that import a String function. 

You'll need to modify the import directive to match other function modules that you import, which is essentially all functions except for the Core and Mule functions, for example, `import * from dw::core::Arrays`. Check here for examples for those modules: https://docs.mulesoft.com/mule-runtime/4.2/dw-functions#dw_modules.

Reply:
Updated. Though, I didn't include `.DataWeave Script` before the examples because it was generating a new line and look weird since we are already explaining the example in the previous sentence.
@afelisatti Could you review the samples and check the function modules for `import * from dw::core::` I was not sure about some of these. So you might leave a comment to have it updated. Thank you.